src/lightning_app/components/training.py

```diff
@@ -0,0 +1,167 @@
+                "PL_TRAINER_STRATEGY": "ddp",
+                "PL_TRAINER_DEVICES": str(self.cloud_compute.devices),
+                "PL_TRAINER_ACCELERATOR": "auto",
+            }
+        )
+        return super().run()
+
+    def on_after_run(self, script_globals):
+        from pytorch_lightning import Trainer
+        from pytorch_lightning.utilities.cli import LightningCLI
+
+        cli = [v for v in script_globals.values() if isinstance(v, LightningCLI)]
+        if cli:
+            trainer = cli[0].trainer
+        else:
+            trainer = [v for v in script_globals.values() if isinstance(v, Trainer)][0]
+
+        if trainer.checkpoint_callback.best_model_score:
+            self.best_model_path = Path(trainer.checkpoint_callback.best_model_path)
+            self.best_model_score = float(trainer.checkpoint_callback.best_model_score)
+        self.has_finished = True
+        # TODO: Why does it hang there.
+        raise SystemExit(0)
+
+    def _trainer_init_pre_middleware(self, trainer, *args, **kwargs):
+        if self.node_rank != 0:
+            return {}, args, kwargs
+
+        from pytorch_lightning.serve import ServableModuleValidator
+
+        callbacks = kwargs.get("callbacks", [])
+        if self.sanity_serving:
+            callbacks = callbacks + [ServableModuleValidator()]
+        kwargs["callbacks"] = callbacks
+        return {}, args, kwargs
+
+    @property
+    def is_running_in_cloud(self) -> bool:
+        return "LIGHTNING_APP_STATE_URL" in os.environ
+
+
+class LightningTrainingComponent(LightningFlow):
+    def __init__(
+        self,
+        script_path: str,
+        script_args: Optional[Union[list, str]] = None,
+        num_nodes: int = 1,
+        cloud_compute: CloudCompute = CloudCompute("default"),
+        sanity_serving: bool = False,
+        script_runner: Type[TracerPythonScript] = PyTorchLightningScriptRunner,
+    ):
+        """This component enables to perform distributed multi-node multi-devices training.
+
+        Example::
+
+            from lightning import LightningApp
+            from lightning.app.components.training import LightningTrainingComponent
+            from lightning.app.utilities.packaging.cloud_compute import CloudCompute
+
+            app = LightningApp(
+                LightningTrainingComponent(
+                    "train.py",
+                    num_nodes=2,
+                    cloud_compute=CloudCompute("gpu"),
+                ),
+            )
+
+        Arguments:
+            script_path: Path to the script to be executed.
+            script_args: The arguments to be pass to the script.
+            num_nodes: Number of nodes.
+            cloud_compute: The cloud compute object used in the cloud.
+            sanity_serving: Whether to validate the model correctly implements
+                the ServableModule API
+        """
+        super().__init__()
+        self.ws = structures.Dict()
```

Comment:
It doesn't look like you use the dictionary keys for anything. Consider making it a list

Reply:
Done !